### ISSUE TITLE
Marathon is Crashing Often

### ISSUE THREAD

**Issue by sethdmoore** [mesosphere/marathon#834]
Hello,

We run a clustered marathon cluster of 5 machines. We've been noticing the marathon leader "switching" every 15m to an hour. This is causing serious problems such as orphaned containers (containers that mesos knows about but marathon does not).

Looking closer into the issue, we've noticed that our marathon masters are crashing in a few different ways

The first way is a segmentation fault. The gist below is the full hs_xxxx.pid segmentation fault log.
[segfault log](https://gist.github.com/sethdmoore/9b1ba7dc63663c7f2843)

The second way it is crashing seems to be related to libmesos
```
F1128 08:41:20.672775  9158 check.hpp:79] Check failed: f.isReady()
*** Check failure stack trace: ***
    @     0x7fcd9e73c4bd  google::LogMessage::Fail()
    @     0x7fcd9e73e2fd  google::LogMessage::SendToLog()
    @     0x7fcd9e73c0ac  google::LogMessage::Flush()
    @     0x7fcd9e73ebf9  google::LogMessageFatal::~LogMessageFatal()
    @     0x7fcd9e7324e4  _checkReady<>()
    @     0x7fcd9e72ff24  Java_org_apache_mesos_state_AbstractState__1_1fetch_1get
    @     0x7fcdb49aef88  (unknown)
```
More log context is available [here](https://gist.github.com/sethdmoore/7c19d2e8620a33cc4080)

The third crash also seems related to libmesos
```
F1128 16:28:24.845263 28864 org_apache_mesos_state_AbstractState.cpp:145] CHECK_READY(*future): is PENDING 
*** Check failure stack trace: ***
    @     0x7f1ccb1664bd  google::LogMessage::Fail()
    @     0x7f1ccb1682fd  google::LogMessage::SendToLog()
    @     0x7f1ccb1660ac  google::LogMessage::Flush()
    @     0x7f1ccb168bf9  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f1ccb159ff7  Java_org_apache_mesos_state_AbstractState__1_1fetch_1get
    @     0x7f1ce167ea88  (unknown)
```
More log context is available [here](https://gist.github.com/sethdmoore/891e4c7deafbc45f617e)

We are running
Zookeeper 3.4.6, 7 nodes
Marathon 0.7.5, 5 nodes
Mesos 0.20.0, 5 nodes

we have somewhere around 60-100 mesos slaves in an autoscaling group connecting to the mesos masters

Java version 6 
```
# java -version
java version "1.6.0_31"
OpenJDK Runtime Environment (IcedTea6 1.13.3) (6b31-1.13.3-1ubuntu1)
OpenJDK 64-Bit Server VM (build 23.25-b01, mixed mode)
```

**Comment by uzyexe** [comment]
Hello, 

I encountered a similar problem. It was encountered when are running about 3000 tasks.

We are running:
Zookeeper 3.4.6, 7 nodes
Marathon 0.7.7-SNAPSHOT, 7 nodes
Mesos 0.21.1, 7 nodes
and, around 25 mesos slaves
java version "1.7.0_65"

A problem could be avoided by adding a JVM option.

```
-d64
-server 
-Xms
-Xms
-XX:NewSize
-XX:MaxNewSize
-XX:PermSize
-XX:MaxPermSize
-XX:+UseConcMarkSweepGC
-XX:+CMSParallelRemarkEnabled
-XX:+UseParNewGC
-XX:SurvivorRatio
-XX:MaxTenuringThreshold
-XX:TargetSurvivorRatio
```

If running about 5000 tasks: 
`-d64 -server -Xms8g -Xmx8g -XX:NewSize=6g -XX:MaxNewSize=6g -XX:PermSize=512m -XX:MaxPermSize=512m -XX:+UseConcMarkSweepGC -XX:+CMSParallelRemarkEnabled -XX:+UseParNewGC -XX:SurvivorRatio=8 -XX:MaxTenuringThreshold=32 -XX:TargetSurvivorRatio=80`

If running about 10000 tasks: 
`-d64 -server -Xms16g -Xmx16g -XX:NewSize=12g -XX:MaxNewSize=12g -XX:PermSize=512m -XX:MaxPermSize=512m -XX:+UseConcMarkSweepGC -XX:+CMSParallelRemarkEnabled -XX:+UseParNewGC -XX:SurvivorRatio=8 -XX:MaxTenuringThreshold=32 -XX:TargetSurvivorRatio=80`

and, adjustment of `--reconciliation_interval` also need.  I'm successful in running about 13000 tasks at present.

I think GC and `state.fetch(id).get(timeout.duration.length, timeout.duration.unit)` has caused problems.
https://github.com/mesosphere/marathon/blob/0.7.7-SNAPSHOT/src/main/scala/mesosphere/marathon/tasks/TaskTracker.scala#L40

**Comment by ConnorDoyle** [comment]
Here's what we know:

1. `org.apache.mesos.state.AbstractState.fetch` returns a Java Future backed by a C++ (libprocess) future.  That Future's `get` method dispatches back into native code, where we see the three failure modes (two kinds of assertion failure, or a segmentation fault.)  See https://github.com/apache/mesos/blob/0.21.1/src/java/src/org/apache/mesos/state/AbstractState.java#L63

1. The Java Futures returned from the Mesos state API define finalizer methods.  These finalizers free the underlying C++ future.  These finalizers should probably be removed.  See https://github.com/apache/mesos/blob/0.21.1/src/java/jni/org_apache_mesos_state_AbstractState.cpp#L226

1. The crashes seem correlated with GC events.  The periodicity of crashes under high load with large heaps seems to correspond with the sawtooth heap allocation graph typical of long-lived JVM processes.

@uzyexe, do you know if the problem is any better or worse with the changes introduced in PR #881?  Based on our current understanding of this problem, we had hoped that it could help.

**Comment by uzyexe** [comment]
@ConnorDoyle, thank you.

I was running marathon v0.7.5, v0.7.6, v0.7.7-SNAPSHOT(Is now v0.8.0-RC1?).
and I also know PR #881. but, It in particular there was no impact in my environment.

**Comment by andrewortman** [comment]
Just ran straight into this issue with 0.7.6 - we plan on upgrading to 0.8.0 this weekend. Happened when we increased our task count significantly. Are there any new updates or workarounds available? 

Thanks

**Comment by kolloch** [comment]
This should be fixed with Marathon 0.8.2-RC3 and Mesos 0.22.1 libraries and master. 

Please reopen if that doesn't fix your problem.
